isis/src/base/objs/CropApp/CropApp.cpp

```diff
@@ -0,0 +1,392 @@
+#include "CropApp.h"
+
+#include <QString>
+#include <QThread>
+#include <QDebug>
+#include <QMetaType>
+
+#include <cmath>
+
+#include "Cube.h"
+#include "ProcessByLine.h"
+#include "SpecialPixel.h"
+#include "LineManager.h"
+#include "FileName.h"
+#include "IException.h"
+#include "Projection.h"
+#include "AlphaCube.h"
+#include "Table.h"
+#include "SubArea.h"
+
+using namespace std;
+using namespace Isis;
+
+
+
+namespace Isis {
+
+
+int cropper::m_sline= -1;
+int cropper::m_linc=-1;
+int cropper::m_sband=-1;
+int cropper::m_ssample=-1;
+int cropper::m_sinc=-1;
+Cube * cropper::m_cube=NULL;
+LineManager * cropper::m_in = NULL;
+
+int cropper::m_osamples =-1;
+int cropper::m_olines=-1;
+int cropper::m_obands=-1;
+
+QString cropper::m_outputCubeName="";
+bool cropper::m_propspice = false;
+
+
+         CropApp::CropApp(QString &from, QString &to, int ssample,int nsamples,
+                          int sinc, int sline,int nlines,int linc,bool propspice){
+
+            m_fromCube =from;
+            m_toCube = to;
+            m_ssample = ssample;
+            m_nsamples = nsamples;
+            m_sinc = sinc;
+            m_sline = sline;
+            m_nlines = nlines;
+            m_linc = linc;
+            m_propspice = propspice;
+            m_sband = 1;
+            m_in = NULL;
+            m_cube = NULL;
+
+            //CubeAttributeInput inAtt(from);
+            //m_cube = new Cube();
+            //m_cube->setVirtualBands(inAtt.bands());
+            //m_cube->open(from);
+
+
+            cropper *crop = new cropper(ssample,nsamples,sinc,
+                                          sline,nlines,linc,propspice,to,from);
+            crop->moveToThread(&cropThread);
+            connect(&cropThread,SIGNAL(started()),crop,
+                    SLOT(cropit()));
+
+            connect(&cropThread, &QThread::finished,crop, &QObject::deleteLater);
+
+            //connect(this, &CropApp::operate, crop, &cropper::cropit);
+            connect(crop, &cropper::resultReady, this, &CropApp::handleResults);
+
+
+         }
+
+         void CropApp::start() {
+
+                 cropThread.start();
+
+
+         }
+
+         void CropApp::handleResults() {
+           qDebug() << "Results are handled here.";
+
+
+         }
+
+
+  cropper::cropper(int ssample,int nsamples,
+          int sinc, int sline,int nlines,int linc,bool propspice,QString to,QString from){
+
+
+
+
+    CubeAttributeInput inAtt(from);
+    m_cube = new Cube();
+    m_cube->setVirtualBands(inAtt.bands());
+    m_cube->open(from);
```

Comment:
You already had to open this cube, It makes more sense to open it once and then just pass it into the cropper object. This also makes this app way more testable.

You can do the same thing with the output cube.

Reply:
Good point.